napari/_qt/qt_resources/_svg.py

```diff
@@ -0,0 +1,149 @@
+"""
+A Class for generating QIcons from SVGs with arbitrary colors at runtime.
+"""
+from __future__ import annotations
+
+from functools import lru_cache
+
+from qtpy.QtCore import QByteArray, QPoint, QRect, QRectF, Qt
+from qtpy.QtGui import QIcon, QIconEngine, QImage, QPainter, QPixmap
+from qtpy.QtSvg import QSvgRenderer
+
+
+class QColoredSVGIcon(QIcon):
+    """A QIcon class that specializes in colorizing SVG files.
+
+    Parameters
+    ----------
+    path_or_xml : str
+        Raw SVG XML or a path to an existing svg file.  (Will raise error on
+        ``__init__`` if a non-existent file is provided.)
+    color : str, optional
+        A valid CSS color string, used to colorize the SVG. by default None.
+    opacity : float, optional
+        Fill opacity for the icon (0-1).  By default 1 (opaque).
+
+    Examples
+    --------
+    >>> from napari._qt.qt_resources import QColoredSVGIcon
+    >>> from qtpy.QtWidgets import QLabel
+
+    # Create icon with specific color
+    >>> label = QLabel()
+    >>> icon = QColoredSVGIcon.from_resources('new_points')
+    >>> label.setPixmap(icon.colored('#0934e2', opacity=0.7).pixmap(300, 300))
+    >>> label.show()
+
+    # Create colored icon using theme
+    >>> label = QLabel()
+    >>> icon = QColoredSVGIcon.from_resources('new_points')
+    >>> label.setPixmap(icon.colored(theme='light').pixmap(300, 300))
+    >>> label.show()
+    """
+
+    def __init__(
+        self, path_or_xml: str, color: str = None, opacity: float = 1.0
+    ) -> None:
+        from ...resources import get_colorized_svg
+
+        self._svg = path_or_xml
+        colorized = get_colorized_svg(path_or_xml, color, opacity)
+        super().__init__(SVGBufferIconEngine(colorized))
+
+    @lru_cache()
+    def colored(
+        self,
+        color: str = None,
+        opacity: float = 1.0,
+        theme: str = None,
+        theme_key: str = 'icon',
+    ):
+        """Return a new colorized QIcon instance.
+
+        Parameters
+        ----------
+        color : str, optional
```

Comment:
How strict are you on `Optional[..]` for typing informations ?

Reply:
I like to be as strict as possible... hadn't looked into this one yet, but I see that explicit optionals is the future, https://github.com/python/mypy/issues/9091, so will add!